source/enum_set.h

```diff
@@ -12,195 +12,249 @@
+    return !(lhs == rhs);
   }
 
-  // Determines whether the given enum value can be represented
-  // as a bit in a uint64_t mask. If so, then returns that mask bit.
-  // Otherwise, returns 0.
-  uint64_t AsMask(uint32_t word) const {
-    if (word > 63) return 0;
-    return uint64_t(1) << word;
+  // Returns true if this set contains at least one value contained in `in_set`.
+  // Note: If `in_set` is empty, this function returns true.
+  bool HasAnyOf(const EnumSet<T>& in_set) const {
+    if (in_set.IsEmpty()) {
+      return true;
+    }
+
+    for (auto& lhs_bucket : buckets) {
+      for (auto& rhs_bucket : in_set.buckets) {
+        if (lhs_bucket.start != rhs_bucket.start) {
+          continue;
+        }
+
+        if (lhs_bucket.data & rhs_bucket.data) {
+          return true;
+        }
+      }
+    }
+    return false;
   }
 
-  // Ensures that overflow_set_ references a set.  A new empty set is
-  // allocated if one doesn't exist yet.  Returns overflow_set_.
-  OverflowSetType& Overflow() {
-    if (overflow_.get() == nullptr) {
-      overflow_ = MakeUnique<OverflowSetType>();
+ private:
+  // Storage for the buckets.
+  std::vector<Bucket> buckets;
+
+  // For a given enum `value`, finds the bucket index that could contain this
+  // value. If no such bucket is found, the index at which the new bucket should
+  // be inserted is returned.
+  size_t find_bucket_for_value(T value) const {
+    // Set is empty, insert at 0.
+    if (buckets.size() == 0) {
+      return 0;
+    }
+
+    size_t index = std::min(buckets.size() - 1, compute_bucket_index(value));
+    const T needle = compute_bucket_start(value);
+
+    const T bucket_start = buckets[index].start;
+    // Computed index is the correct one.
+    if (bucket_start == needle) {
+      return index;
     }
-    return *overflow_;
+
+    // Bucket contains smaller values. Linear scan right.
+    if (bucket_start < needle) {
+      for (index += 1; index < buckets.size() && buckets[index].start < needle;
+           index++) {
+      }
+      return index;
+    }
+
+    // Bucket contains larger values, insert front.
+    if (index == 0) {
+      return index;
+    }
+
+    for (index -= 1; index > 0 && buckets[index].start > needle; index--) {
+    }
+    return buckets[index].start >= needle ? index : index + 1;
   }
 
-  // Enums with values up to 63 are stored as bits in this mask.
-  uint64_t mask_ = 0;
-  // Enums with values larger than 63 are stored in this set.
-  // This set should normally be empty or very small.
-  std::unique_ptr<OverflowSetType> overflow_ = {};
+  // Creates a new bucket to store `value` and inserts it at `index`.
+  // If the `index` is past the end, the bucket is inserted at the end of the
+  // vector.
+  void create_bucket_for(size_t index, T value) {
+    Bucket bucket = {1ULL << compute_bucket_offset(value),
+                     compute_bucket_start(value)};
+    buckets.emplace(buckets.begin() + index, std::move(bucket));
```

Comment:
This causes insertions to be linear time.

Reply:
Depends of the insertion pattern: if enums are added in order of indices, it's a push_back() every 64 enums.
But the vector has to be sorted, so yes, if we insert in reverse order, we got a move.